pandas/tests/io/parser/test_parse_dates.py

```diff
@@ -848,3 +860,80 @@ def test_parse_timezone(all_parsers):
 
     expected = DataFrame(expected_data)
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("date_string", [
+    "32/32/2019",
+    "02/30/2019",
+    "13/13/2019",
+    "13/2019",
+    "a3/11/2018",
+    "10/11/2o17"
+])
+def test_invalid_parse_delimited_date(all_parsers, date_string):
+    parser = all_parsers
+    expected = DataFrame({0: [date_string]}, dtype="object")
+    result = parser.read_csv(StringIO(date_string),
+                             header=None, parse_dates=[0])
+    tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("date_string,dayfirst,expected", [
+    # %d/%m/%Y; month > 12 thus replacement
+    ("13/02/2019", False, datetime(2019, 2, 13)),
+    ("13/02/2019", True, datetime(2019, 2, 13)),
+    # %m/%d/%Y; day > 12 thus there will be no replacement
+    ("02/13/2019", False, datetime(2019, 2, 13)),
+    ("02/13/2019", True, datetime(2019, 2, 13)),
+    # %d/%m/%Y; dayfirst==True thus replacement
+    ("04/02/2019", True, datetime(2019, 2, 4))
+])
+def test_parse_delimited_date_swap(all_parsers, date_string,
+                                   dayfirst, expected):
+    parser = all_parsers
+    expected = DataFrame({0: [expected]}, dtype="datetime64[ns]")
+    result = parser.read_csv(StringIO(date_string), header=None,
+                             dayfirst=dayfirst, parse_dates=[0])
+    tm.assert_frame_equal(result, expected)
+
+
+def _helper_hypothesis_delimited_date(call, date_string, **kwargs):
+    msg, result = None, None
+    try:
+        result = call(date_string, **kwargs)
+    except ValueError as er:
+        msg = str(er)
+        pass
+    return msg, result
+
+
+@given(gen_random_datetime)
+@pytest.mark.parametrize("delimiter", list(" -./"))
+@pytest.mark.parametrize("dayfirst", [True, False])
+@pytest.mark.parametrize("date_format", [
+    "%d %m %Y",
+    "%m %d %Y",
+    "%m %Y",
+    "%Y %m %d",
+    "%y %m %d",
+    "%Y%m%d",
+    "%y%m%d",
+])
+def test_hypothesis_delimited_date(date_format, dayfirst, delimiter, date):
+    if date_format == "%m %Y" and delimiter == ".":
+        pytest.skip("parse_datetime_string cannot reliably tell whether \
+        e.g. %m.%Y is a float or a date, thus we skip it")
+    result, expected = None, None
+    except_in_dateutil, except_out_dateutil = None, None
+    date_string = date.strftime(date_format.replace(' ', delimiter))
+
+    except_out_dateutil, result = _helper_hypothesis_delimited_date(
+        parse_datetime_string, date_string,
```

Comment:
Not quite right. We call the `parse_datetime_string` (not written by us), which in turn calls our `_parse_delimited_date`. Therefore, this test must pass before our changes.

As for hitting the module specific code, in the situation described by this expression: `if date_format == "% m% Y" and delimiter == "."` we hit the function inside the `parse_datetime_string`, which makes it's behavior different from the `du_parse` behavior. But for now this is the best solution we found.

Reply:
@anmyachev @pganssle has a (totally reasonable) habit/policy in dateutil of avoiding testing internal functions, but that is not shared in pandas.  We definitely do need to test the user-facing parsing functions (`to_datetime`, `Timestamp`), but it is also very much OK to directly test the lower-level implementations.  That way if the high-level test starts breaking, it is easier to track down the problem.

Reply:
@jbrockmendel glad to hear it! Then did I understand correctly that there were enough tests?

Reply:
To clarify, while I do think that is a wise *general* policy, this is not a matter of policy but about what precisely is being tested. In this case, you have a performance-only fix, so the ideal situation is to show that the automated tests explore enough of the test space that you can be somewhat confident that it will not introduce regressions.

The best way to do this is to put regression tests in place that will pass both *before* and *after* the changes (thus making it fairly easy to show that no new behavior was introduced), and to test it at the level of the public interface - which is the part that might actually introduce regressions. I don't know enough about the code to know that your tests are adequately showing that this new function won't cause any performance differences in the higher level code, and that's part of the problem - by testing the lower-level stuff you may be testing it in an entirely "unnatural" context that gets you good coverage metrics, but doesn't mimic real-world usage.

That said, my objection was mainly around the *first* case because when I searched for `parse_datetime_string` it seemed like that was a function you added. If @jbrockmendel and the other maintainers feel that this is the right level for the tests (and "we used to call dateutil here, now we call `_parse_delimited_date`, therefore as long as `_parse_delimited_date` does whatever dateutil did, it's fine" is a very reasonable test strategy), I don't need to press the point.

Reply:
@jbrockmendel @jreback what do you think about https://github.com/pandas-dev/pandas/pull/25922#discussion_r274042241?

What our next step?

Reply:
ping? what else you guys feel we need to do to make this acceptable?